src/EventStore.Core.Tests/Services/GossipService/NodeGossipServiceTests.cs

```diff
@@ -338,7 +348,463 @@ public class when_gossip_received_with_older_timestamp_about_peer_node :
+					MemberInfoForVNode(_nodeThree, _timeProvider.UtcNow, epochNumber: 1, esVersion: "1.1.1.3")),
+				_nodeTwo.HttpEndPoint));
+
+		protected override Message When() =>
+			new GossipMessage.ClientGossip(new CallbackEnvelope(CaptureGossipReply));
+
+		private ClientClusterInfo GetExpectedClusterInfo() {
+			return new ClientClusterInfo(new ClusterInfo(
+				MemberInfoForVNode(_currentNode, _timeProvider.UtcNow, esVersion: VersionInfo.DefaultVersion),
+				MemberInfoForVNode(_nodeTwo, _timeProvider.UtcNow, epochNumber: 1, esVersion: "1.1.1.2"),
+				MemberInfoForVNode(_nodeThree, _timeProvider.UtcNow, epochNumber: 1, esVersion: "1.1.1.3")), _currentNode.HttpEndPoint.GetHost(), _currentNode.HttpEndPoint.GetPort());
+		}
+		
+		[Test]
+		public void reply_should_have_version_info() {
+			AssertEx.AssertUsingDeepCompare(_capturedMessage,
+				new GossipMessage.SendClientGossip(GetExpectedClusterInfo()));
+		}
+
+		private void CaptureGossipReply(Message message) => _capturedMessage = message;
+	}
+
+	public class if_gossip_received_has_alive_nodes_on_old_version : NodeGossipServiceTestFixture {
+		protected override Message[] Given() => GivenSystemInitializedWithKnownGossipSeedSources();
+		protected override Message When() =>
+			new GossipMessage.GossipReceived(new CallbackEnvelope(AssertGossipReply),
+				new ClusterInfo(MemberInfoForVNode(_nodeTwo, _timeProvider.UtcNow, epochNumber: 1, esVersion: VersionInfo.DefaultVersion),
+					// nodeThree is dead
+					MemberInfoForVNode(_nodeThree, _timeProvider.UtcNow, epochNumber: 1, esVersion: "1.1.1.3",
+						isAlive: false),
+					MemberInfoForVNode(_nodeFour, _timeProvider.UtcNow, epochNumber: 1, esVersion: VersionInfo.OldVersion)), _nodeTwo.HttpEndPoint);
+		
+		private ClusterInfo GetExpectedClusterInfo() {
+			return new ClusterInfo(
+				MemberInfoForVNode(_currentNode, _timeProvider.UtcNow, esVersion: VersionInfo.DefaultVersion),
+				MemberInfoForVNode(_nodeTwo, _timeProvider.UtcNow, epochNumber: 1, esVersion: VersionInfo.DefaultVersion),
+				MemberInfoForVNode(_nodeThree, _timeProvider.UtcNow, epochNumber: 1, esVersion: "1.1.1.3", isAlive: false),
+				MemberInfoForVNode(_nodeFour, _timeProvider.UtcNow, epochNumber: 1, esVersion: VersionInfo.OldVersion));
+		}
+
+		private Dictionary<EndPoint, string> GetExpectedEndPointVsVersion() {
+			Dictionary<EndPoint, string> endpointVsVersion = new Dictionary<EndPoint, string> {
+				{ _nodeFour.HttpEndPoint, VersionInfo.OldVersion },
+				{ _nodeTwo.HttpEndPoint, VersionInfo.DefaultVersion },
+				{ _currentNode.HttpEndPoint, VersionInfo.DefaultVersion }
+			};
+			return endpointVsVersion;
+		}
+
+		[Test]
+		public void should_detect_version_mismatch() {
+			ExpectMessages(new GossipMessage.GossipUpdated(GetExpectedClusterInfo()));
+			
+			Dictionary<EndPoint, string> ipAddressVsVersion = ClusterStateChangeListener.GetIPAddressVsVersion(GetExpectedClusterInfo(), out int numDistinctKnownVersions);
```

Comment:
Nothing mayor, but would be nice to have these in a separate file same.

Reply:
Makes sense. I have created a separate test file for these tests.